src/index/indexer.c

```diff
@@ -0,0 +1,257 @@
+				// decrease graph reference count
+				GraphContext_DecreaseRefCount(ctx.gc);
+				break;
+			case INDEXER_DROP:
+				rm_ctx = RedisModule_GetThreadSafeContext(NULL);
+				RedisModule_ThreadSafeContextLock(rm_ctx);
+
+				Index_Free(ctx.idx);
+
+				RedisModule_ThreadSafeContextUnlock(rm_ctx);
+				break;
+			default:
+				assert(false && "unknown indexer operation");
+				break;
+		}
+	}
+
+	return NULL;
+}
+
+// add task to indexer queue
+static void _indexer_AddTask
+(
+	IndexPopulateCtx *task  // task to be added
+) {
+	// lock
+	int res = pthread_mutex_lock(&indexer->m);
+	ASSERT(res == 0);
+
+	// add task to queue
+	res = CircularBuffer_Add(indexer->q, task);
+	ASSERT(res == 1);
+
+	// unlock
+	res = pthread_mutex_unlock(&indexer->m);
+	ASSERT(res == 0);
+
+	// signal conditional variable
+	pthread_mutex_lock(&indexer->cm); 
+	pthread_cond_signal(&indexer->c);
+	pthread_mutex_unlock(&indexer->cm);
+}
+
+// pops a task from queue
+// if queue is empty caller will be waiting on conditional variable
+static void _indexer_PopTask
+(
+	IndexPopulateCtx *task
+) {
+	ASSERT(task != NULL);
+
+	// lock queue
+	int res = pthread_mutex_lock(&indexer->m);
+	ASSERT(res == 0);
+
+	// remove task to queue
+	if(CircularBuffer_Empty(indexer->q)) {
+		// waiting for work
+		// lock conditional variable mutex
+		pthread_mutex_lock(&indexer->cm);
+
+		// unlock queue mutex
+		pthread_mutex_unlock(&indexer->m);
+
+		// wait on conditional variable
+		pthread_cond_wait(&indexer->c, &indexer->cm);
+		pthread_mutex_unlock(&indexer->cm);
+
+		// work been added to queue
+		// lock queue
+		int res = pthread_mutex_lock(&indexer->m);
+		ASSERT(res == 0);
+	}
+
+	res = CircularBuffer_Remove(indexer->q, task);
+	ASSERT(res == 1);
+
+	// unlock
+	res = pthread_mutex_unlock(&indexer->m);
+	ASSERT(res == 0);
+}
+
+// initialize indexer
+// create indexer worker thread and task queue
+bool Indexer_Init(void) {
+	ASSERT(indexer == NULL);
+
+	int a_res  = 0;  // attribute create result code
+	int c_res  = 0;  // conditional variable create result code
+	int t_res  = 0;  // thread create result code
+	int m_res  = 0;  // mutex create result code
+	int cm_res = 0;  // conditional variable mutex create result code
+
+	indexer = rm_calloc(1, sizeof(Indexer));
+
+	// create queue mutex
+	m_res = pthread_mutex_init(&indexer->m, NULL);
+	if(m_res != 0) {
+		goto cleanup;
+	}
+
+	// create conditional variable mutex
+	cm_res = pthread_mutex_init(&indexer->cm, NULL);
+	if(cm_res != 0) {
+		goto cleanup;
+	}
+
+	// create conditional var
+	c_res = pthread_cond_init(&indexer->c, NULL);
+	if(c_res != 0) {
+		goto cleanup;
+	}
+
+	// create task queue
+	indexer->q = CircularBuffer_New(sizeof(IndexPopulateCtx), 256);
+
+	// create worker thread
+	pthread_attr_t attr;
+	a_res = pthread_attr_init(&attr);
+	a_res = pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+	if(a_res != 0) {
+		goto cleanup;
+	}
+
+    t_res = pthread_create(&indexer->t, &attr, _index_populate, NULL);
```

Comment:
indentation

Reply:
could be a GitHub issue, seems to be indented on my editor